Add documentation for using Tekton with Prow 🙌

bobcatfish opened the pull request:
This commit adds the beginning of some usage documentation about how to
get Tekton and Prow working. I think there's probably more to add but I hope this can
be a good starting point!

Please let me know if the docs or the examples should be in different locations, I basically just tried to guess and pick something but am happy to move things around!

I wouldn't have been able to do this without @JoelSpeed's guidance in
https://github.com/kubernetes/test-infra/issues/13874#issuecomment-521237150
Thanks @JoelSpeed!!! ❤️

I was able to follow along with these docs and get Tekton + Prow running
for `tektoncd` itself: https://github.com/tektoncd/plumbing/pull/66 🎉

Fixes #13874 (and maybe https://github.com/tektoncd/pipeline/issues/1187 as well?)

bobcatfish commented:
> Keywords which can automatically close issues and at(@) mentions are not allowed in commit messages.

I understand issue closing but I am curious why @-ing is not okay :O

stevekuznetsov commented:
Whoever is on the other side of the at-mention will get a notification every time:

 - you push the commit
 - your commit is merged into a branch, including:
  a. when this PR merges
  b. when the commit is cherry-picked into other branches
  c. when commits move around in forks of the repo

It ends up being pretty spammy :)

For review
/assign @cjwagner @ccojocar

ccojocar commented on prow/cluster/simple_pipeline_deployment.yaml:
```diff
@@ -0,0 +1,33 @@
+kind: Deployment
```
There is already a deployment in `pipeline_deployment.yaml`. I am wondering wha's the reason to not reuse it.

cjwagner replied:
That flag is for running the jobs in a different cluster.

I think it is too misleading to include these deployment files in the `prow/cluster` directory. With the exception of `starter.yaml` (which we should probably move) all of the files under `prow/cluster` are deployed to the prow.k8s.io instance, but the files you've added wouldn't work for our instance.

It looks like this is a more detailed version of `prow/cmd/pipeline/dev.yaml`. Consider updating that file or adding this one to that directory.

ccojocar replied:
I see, it can be actually pointed to the same cluster. It's true there is an additional configuration step to set up that secret. 

Maybe it shouldn't fail with an error if the provided file is empty/missing. Sounds like a warnings message and use the same cluster is more appropriate in that case.

bobcatfish replied:
> It looks like this is a more detailed version of prow/cmd/pipeline/dev.yaml. Consider updating that file or adding this one to that directory.

Ah I didn't even see dev.yaml! For now I've moved these examples into `cmd/pipeline` and `cmd/crier`

ccojocar commented on prow/cluster/simple_crier_deployment.yaml:
```diff
@@ -0,0 +1,40 @@
+apiVersion: apps/v1
+kind: Deployment
```
There is a deployment for crier in `crier_deployment.yaml`. I am wondering what's the reason to not reuse it.

bobcatfish replied:
Similar to above comment about pipeline_deployment.yaml, https://github.com/kubernetes/test-infra/blob/master/prow/cluster/crier_deployment.yaml includes an optional `job-config` configmap

cjwagner replied:
As with the pipeline controller, this change would prevent crier from working properly on out instance so including it in `prow/cluster` is kind of misleading.

If you want to add a set of deployment files for a Prow instance that only uses tekton, consider grouping those in a separate directory to make it clear what the purpose is and so that users know what components they actually need to deploy (e.g. sinker and at least one of `hook`, `horologium`, or the `gerrit` adapter will be needed to actually trigger jobs.)

cjwagner replied:
Ah, I see that in the docs you have the user first follow `getting_started_deploy.md` so `hook`, `horologium`, etc will already be deployed.

ccojocar commented on prow/tekton.md:
````diff
@@ -0,0 +1,142 @@
+
+```
+
+### Install Tekton Pipelines
````
I am not sure but I think this step needs to be before installing the controller because the controller expects the PipelineRun CRD to be already installed in order to be able to start watching.

bobcatfish replied:
kk i can try to make that more clear

ccojocar commented on prow/tekton.md:
```diff
@@ -0,0 +1,142 @@
+      trigger: # Required by versions 0.2.0 - v0.3.1 of Tekton Pipelines
+        type: manual
+      pipelineRef:
+        name: special-pipeline # Use the Tekton Pipeline called `special-pipeline`
```
I think the `serviceAccount` needs also to be defined to use the Tekton's service account instead of `default`. That's the service account with the GitHub token linked to it.

bobcatfish replied:
>  It should be the Tekton's service account with the GitHub token linked to it.

Maybe you can explain more about what this service account would be doing? You can [optionally provide a serviceAccount to a PipelineRun](https://github.com/tektoncd/pipeline/blob/master/docs/pipelineruns.md#service-account) but that just makes it available to executing Tasks if they need it (e.g. to push to an image registry).

Re the GitHub token, if the PipelineRun was trying to push something to GitHub for example, if you wanted [to use the PullRequest resource](https://github.com/tektoncd/pipeline/blob/master/docs/resources.md#pull-request-resource) and push to GitHub from the Pipeline itself, you'd need to provide a secret in the `PipelineResource`

TL;DR it depends on what the PipelineRun is doing - I can add a link to the PipelineRun docs and try to make that more clear here.

> I am not sure that service account is used by default by the Tekton controller.

The controller uses the [` tekton-pipelines-controller` service account](https://github.com/tektoncd/pipeline/blob/master/config/200-serviceaccount.yaml) by default (https://github.com/tektoncd/pipeline/blob/master/config/controller.yaml#L28) - but pods that execute as part of a PipelineRun would use the default service account of the namespace by default, unless `serviceAccount` or `serviceAccounts` are specified

bobcatfish replied:
I've added a link to the PipelineRun docs right below this and mentioned ServiceAccounts!

ccojocar replied:
I mean this service account https://github.com/tektoncd/pipeline/blob/master/docs/auth.md#basic-authentication-git, but pointing to the doc is enough if someone needs to configure it. Thanks for adding that!

bobcatfish commented:
I made some changes, lemme know what you think @ccojocar ! Also let me know what you want to do about the differences between those example yaml files :S

cjwagner commented on prow/tekton.md:
```diff
@@ -0,0 +1,147 @@
+**Warning**
+
+Prow users may expect to see logs from their `PipelineRuns` be visible in
+Gubernator, however this functionality is not supported yet, and logs must be
```
Gubernator is deprecated in favor of Spyglass.  It might be more accurate to note that pipeline prowjobs cannot be decorated with the pod utilities so job logs, metadata, and artifacts are not uploaded to GCS.

bobcatfish replied:
kk updated!

cjwagner commented on prow/tekton.md:
```diff
@@ -0,0 +1,147 @@
+    skip_report: false
+    pipeline_run_spec: # Template for creation of the `PipelineRun`
+      trigger: # Required by versions 0.2.0 - v0.3.1 of Tekton Pipelines
+        type: manual
```
If this is required should we have the pipeline prow controller add it if it isn't present?

bobcatfish replied:
that would work! even better tho would be to update the version of Pipelines that Prow is using tho - this field is deprecated

cjwagner commented on prow/tekton.md:
```diff
@@ -0,0 +1,147 @@
+    rerun_command: "/run do-the-pipeline"
+    trigger: "(?m)^/run do-the-pipeline,?(\\s+|$)"
+    always_run: true
+    skip_report: false
```
nit: I'd omit the `skip_report`, `rerun_command`, and `trigger` fields. They aren't necessary if you are ok with using `/test do-the-pipeline` to trigger the job.

bobcatfish replied:
kk!

cjwagner commented on prow/tekton.md:
```diff
@@ -0,0 +1,147 @@
+* `pipelineRef` - Refers to a
+  [`Pipeline`](#create-and-apply-pipelines-and-tasks) that must exist for the
+  `PipelineRun` to be able to execute
+* [`PROW_IMPLICIT_GIT_REF`] - Using this special
```
It is probably worth noting that we also support `PROW_EXTRA_GIT_REF_0`, `PROW_EXTRA_GIT_REF_1`, etc. if the ProwJob specifies `extraRefs` e.g. https://github.com/kubernetes/test-infra/blob/6eb5baa91054e17e8c1db7f7cda3537170f6a1be/config/jobs/kubernetes/test-infra/test-infra-trusted.yaml#L677-L680

bobcatfish replied:
thanks for mentioning this, this saved my life when I went to use this for periodic jobs XD

cjwagner commented on prow/tekton.md:
```diff
@@ -0,0 +1,147 @@
+1. [Setting up Pipelines with Prow](#setting-up-pipelines-with-prow)
+2. [Configuring a `tekton-pipeline` job](#configuring-a-tekton-pipeline-job)
+
+**Warning**
```
The other caveat is that git pipelineresources are currently not garbage collected. (I think we can fix this by updating the pipelineresources with owner references after the pipelinerun is created.)

ccojocar replied:
I implemented this initially in the controller and it worked well, but I was asked to remove it due to the multi-cluster build support. Referencing the `ProwJob` is problematic when the builds are executed in a different cluster but I don't see an issue with a reference to the `PipelineRun`.

bobcatfish replied:
Added a warning about this! Btw https://github.com/tektoncd/pipeline/issues/544 is about this and it looks like this will be addressed in https://github.com/tektoncd/pipeline/issues/872 (which was your request originally @cjwagner!) - you'd have to update Prow to be able to embed PipelineResources tho :)

bobcatfish commented:
Quick update: thanks for the detailed review @cjwagner + @ccojocar - I'm planning to come back and update the PR with your suggestions just haven't gotten a chance yet :pray:

bobcatfish commented:
Okay @cjwagner @ccojocar I finally came back to this (sorry for the delay! :pray: ) and I think I've addressed all your feedback so far (I also added a periodic example), PTAL :D

matthyx commented:
/ok-to-test

cjwagner commented on prow/cmd/crier/simple_crier_deployment.yaml:
```diff
@@ -0,0 +1,40 @@
+        - --github-workers=1
+        - --config-path=/etc/config/config.yaml
+        - --github-endpoint=http://ghproxy
+        - --github-endpoint=https://api.github.com
```
The instructions you provided don't include setting up ghproxy so the `--github-endpoint` flags will make every request fail on the first attempt. I'd suggest removing them both.

cjwagner commented on prow/cmd/crier/simple_crier_deployment.yaml:
```diff
@@ -0,0 +1,40 @@
+      labels:
+        app: crier
+    spec:
+      # Using the service account defined in prow/cluster/pipeline_rbac.yaml
```
```suggestion
      # Using the service account defined in prow/cluster/crier_rbac.yaml
```

cjwagner commented on prow/tekton.md:
```diff
@@ -0,0 +1,172 @@
+## Setting up Pipelines with Prow
+
+1. [Setup and configure Prow itself](getting_started_deploy.md)
+2. [Install Tekton Pipelines](#install-tekton)
```
```suggestion
2. [Install Tekton Pipelines](#install-tekton-pipelines)
```

matthyx commented:
@bobcatfish will you have time to address these nits? It's too bad to delay merging for small stuff...

zzq889 commented on prow/tekton.md:
```diff
@@ -0,0 +1,172 @@
+can start, and in order to be able to execute [Tekton Pipelines](https://github.com/tektoncd/pipeline,
+you also need to install and setup Tekton Pipelines itself.
+
+[Prow is currently compatible with versions 0.2.0 - 0.3.1 of Tekton Pipelines](https://github.com/kubernetes/test-infra/issues/13948)
```
0.8.0 are supported now

bobcatfish commented:
Just in case I ever get back to this, prow supports tekton pipelines 0.10.1 now https://github.com/kubernetes/test-infra/pull/16099

(sorry @matthyx !!! im gonna try to get this in asap 😭  totally happy if someone else is able to take this over tho!)

bobcatfish commented:
At this rate I don't think I'm getting back to this anytime soon :( Hopefully what I wrote can still be helpful if someone else can take this on later!
